src/main/scala/analysis/Cfg.scala

```diff
@@ -51,382 +196,930 @@ trait CfgNode:
+  */
+case class CfgStatementNode(
+    override val id: Int = CfgNode.nextId(),
+    override val predIntra: mutable.Set[CfgEdge] = mutable.Set[CfgEdge](),
+    override val predInter: mutable.Set[CfgEdge] = mutable.Set[CfgEdge](),
+    override val succIntra: mutable.Set[CfgEdge] = mutable.Set[CfgEdge](),
+    override val succInter: mutable.Set[CfgEdge] = mutable.Set[CfgEdge](),
+    data: Statement,
+    block: Block
+) extends CfgCommandNode:
   override def toString: String = s"[Stmt] $data"
 
-/** A control-flow graph.
- * @param entry
- *   the entry to the graph (source)
- * @param exits
- *   the exit to he graph (sink)
- */
-class Cfg(cfg: Option[Cfg] = None):
-  var nodeToBlock: mutable.Map[CfgNode, Block] = mutable.Map()
-  var edges: ListBuffer[Edge] = ListBuffer()
-  var entries: ListBuffer[CfgNode] = ListBuffer()
-  cfg match {
-    case Some(cfg) =>
-      this.edges = cfg.edges
-      this.entries = cfg.entries
-      this.nodeToBlock = cfg.nodeToBlock
-    case _ =>
+  /** Copy this node, but give unique ID and reset edges */
+  override def copyNode(): CfgStatementNode = CfgStatementNode(data = this.data, block = this.block)
+
+/** CFG's representation of a jump. This is used as a general jump node, for both indirect and direct calls.
+  */
+case class CfgJumpNode(
+    override val id: Int = CfgNode.nextId(),
+    override val predIntra: mutable.Set[CfgEdge] = mutable.Set[CfgEdge](),
+    override val predInter: mutable.Set[CfgEdge] = mutable.Set[CfgEdge](),
+    override val succIntra: mutable.Set[CfgEdge] = mutable.Set[CfgEdge](),
+    override val succInter: mutable.Set[CfgEdge] = mutable.Set[CfgEdge](),
+    data: Jump,
+    block: Block
+) extends CfgCommandNode:
+  override def toString: String = s"[Jmp] $data"
+
+  /** Copy this node, but give unique ID and reset edges */
+  override def copyNode(): CfgJumpNode = CfgJumpNode(data = this.data, block = this.block)
+
+/** A general purpose node which in terms of the IR has no functionality, but can have purpose in the CFG. As example,
+  * this is used as a "block" start node for the case that a block contains no statements, but has a `GoTo` as its jump.
+  * In this case we introduce a ghost node as the start of the block for the case that some part of the program jumps
+  * back to this conditional jump (e.g. in the case of loops).
+  */
+case class CfgGhostNode(
+    override val id: Int = CfgNode.nextId(),
+    override val predIntra: mutable.Set[CfgEdge] = mutable.Set[CfgEdge](),
+    override val predInter: mutable.Set[CfgEdge] = mutable.Set[CfgEdge](),
+    override val succIntra: mutable.Set[CfgEdge] = mutable.Set[CfgEdge](),
+    override val succInter: mutable.Set[CfgEdge] = mutable.Set[CfgEdge](),
+    data: NOP = NOP(),
+    block: Block
+) extends CfgCommandNode:
+  override def toString: String = s"[NOP]"
+
+  /** Copy this node, but give unique ID and reset edges */
+  override def copyNode(): CfgGhostNode = CfgGhostNode(block = this.block)
```

Comment:
This isn't actually used for anything is it? Do we need this?

Reply:
@yousifpatti

Reply:
I do not think this is needed as blocks without stmts is being dealt with without adding intermediate nodes